File: indexter-games/src/main/java/com/github/ykrasik/indexter/games/datamodel/persistence/Game.java
package com.github.ykrasik.indexter.games.datamodel.persistence;

import com.github.ykrasik.indexter.games.datamodel.GamePlatform;
import com.github.ykrasik.indexter.games.datamodel.ImageData;
import com.github.ykrasik.indexter.games.datamodel.info.GameInfo;
import com.github.ykrasik.indexter.id.Id;
import lombok.EqualsAndHashCode;
import lombok.NonNull;
import lombok.ToString;
import lombok.Value;
import lombok.experimental.Builder;

import java.nio.file.Path;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

/**
 * @author Yevgeny Krasik
 */
@Value
@Builder
@EqualsAndHashCode(of = "id")
@ToString(of = {"id", "name", "path"})
public class Game {
    @NonNull private final Id<Game> id;
    @NonNull private final Path path;

    @NonNull private final String name;
    @NonNull private final GamePlatform platform;
    @NonNull private final Optional<String> description;
    @NonNull private final Optional<LocalDate> releaseDate;
    @NonNull private final Optional<Double> criticScore;
    @NonNull private final Optional<Double> userScore;
    @NonNull private final Optional<String> giantBombApiDetailsUrl;
    @NonNull private final String url;

    @NonNull private final Optional<ImageData> thumbnail;
    @NonNull private final Optional<ImageData> poster;

    @NonNull private final LocalDateTime lastModified;

    @NonNull private final List<Genre> genres;

    public static Game from(Id<Game> id,
                            Path path,
                            GamePlatform platform,
                            LocalDateTime lastModified,
                            GameInfo info,
                            List<Genre> genres) {
        return builder()
            .id(id)
            .path(path)
            .platform(platform)
            .name(info.getName())
            .lastModified(lastModified)
            .description(info.getDescription())
            .releaseDate(info.getReleaseDate())
            .criticScore(info.getCriticScore())
            .userScore(info.getUserScore())
            .giantBombApiDetailsUrl(Optional.of(info.getDetailUrl()))
            .url(info.getUrl())
            .thumbnail(info.getThumbnail())
            .poster(info.getPoster())
            .genres(genres)
            .build();
    }
}

File: gamedex-common/src/main/kotlin/com/github/ykrasik/gamedex/common/datamodel/Game.kt
package com.github.ykrasik.gamedex.common.datamodel

import org.joda.time.DateTime
import org.joda.time.LocalDate
import java.io.File

/**
 * User: ykrasik
 * Date: 26/05/2016
 * Time: 15:32
 */
data class Game(
    val id: Int,
    private val metaData: MetaData,
    private val gameData: GameData,
    val providerData: List<ProviderData>
) {
    val libraryId: Int get() = metaData.libraryId
    val path: File get() = metaData.path
    val lastModified: DateTime get() = metaData.lastModified

    val name: String get() = gameData.name
    val description: String? get() = gameData.description
    val releaseDate: LocalDate? get() = gameData.releaseDate

    val criticScore: Double? get() = gameData.criticScore
    val userScore: Double? get() = gameData.userScore

    val genres: List<String> get() = gameData.genres

    override fun toString() = "Game(id = $id, name = $name, path = $path)"
}

data class MetaData(
    val libraryId: Int,
    val path: File,
    val lastModified: DateTime
)

data class GameData(
    val name: String,
    val description: String?,
    val releaseDate: LocalDate?,

    val criticScore: Double?,
    val userScore: Double?,

    val genres: List<String>
)

data class ProviderData(
    val type: DataProviderType,
    val apiUrl: String,
    val url: String
)

data class ImageData(
    val thumbnailUrl: String?,
    val posterUrl: String?,
    val screenshot1Url: String?,
    val screenshot2Url: String?,
    val screenshot3Url: String?,
    val screenshot4Url: String?,
    val screenshot5Url: String?,
    val screenshot6Url: String?,
    val screenshot7Url: String?,
    val screenshot8Url: String?,
    val screenshot9Url: String?,
    val screenshot10Url: String?
)